File: src/main/scala/dfops.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import java.io.File
import org.apache.spark.ml.feature.StringIndexer
import org.apache.spark.ml.feature.VectorAssembler
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.sql.Row


class FireSchema{
    val schema = StructType(
        Array(
            StructField("CallNumber", IntegerType, true),
            StructField("UnitID", StringType, true),
            StructField("IncidentNumber", IntegerType, true),
            StructField("CallType", StringType, true),
            StructField("CallDate", StringType, true),
            StructField("WatchDate", StringType, true),
            StructField("CallFinalDisposition", StringType, true),
            StructField("AvailableDtTm", StringType, true),
            StructField("Address", StringType, true),
            StructField("City", StringType, true),
            StructField("Zipcode", IntegerType, true),
            StructField("Battalion", StringType, true),
            StructField("StationArea", StringType, true),
            StructField("Box", StringType, true),
            StructField("OriginalPriority", StringType, true),
            StructField("Priority", StringType, true),
            StructField("FinalPriority", IntegerType, true),
            StructField("ALSUnit", BooleanType, true),
            StructField("CallTypeGroup", StringType, true),
            StructField("NumAlarms", IntegerType, true),
            StructField("UnitType", StringType, true),
            StructField("UnitSequenceInCallDispatch", IntegerType, true),
            StructField("FirePreventionDistrict", StringType, true),
            StructField("SupervisorDistrict", StringType, true),
            StructField("Neighborhood", StringType, true),
            StructField("Location", StringType, true),
            StructField("RowID", StringType, true),
            StructField("Delay", FloatType, true)
        )
    )
}

object FireOps{
    def main(args: Array[String]): Unit ={
        
        // if (args.length <= 0){
        //     println("usage FireOps <File fire dataset>")
        //     System.exit(1)
        // }

        val spark = SparkSession.builder.appName("FireReader").getOrCreate()


        // val fireFile = args(1)

        val schema = new FireSchema().schema

        // reading file 
        val fileLoc = "/app/src/main/data/sf-fire-calls.csv"

        val fireDF = spark.read.schema(schema).option("header", true).csv(fileLoc)

        fireDF.show(numRows = 10)
        // val parquetPath = "/app/src/main/data/scala/parquet/fire"
    
        // createLocalFolderIfNotExists(parquetPath)

        // saving parquet file
        // fireDF.write.mode("overwrite").format("parquet").save(parquetPath)

        val fewFireDF = fireDF
        .select("IncidentNumber", "AvailableDtTm", "CallType")
        .where(col("CallType") =!= "Medical Incident")

        fewFireDF.show(numRows = 5, truncate = false)

        val distinctFireCalls = fireDF
        .select("CallType")
        .where(col("CallType").isNotNull)
        .agg(countDistinct(col("CallType").alias("DistinctCallTypes")))
        distinctFireCalls.show()

        // renaming columns
        val renameCol = fireDF
        .withColumnRenamed("Delay", "ResponseDelayinMins")
        .select("ResponseDelayinMins")
        .where(col("ResponseDelayinMins") > 5)


        val fireTsDF = fireDF
        .withColumn("IncidentDate", to_timestamp(col("CallDate"), "MM/dd/yyyy"))
        .drop("CallDate")
        .withColumn("OnWatchDate", to_timestamp(col("WatchDate"), "MM/dd/yyyy"))
        .drop("WatchDate")
        .withColumn("AvailableDtTs", to_timestamp(col("AvailableDtTm"), "MM/dd/yyyy hh:mm:ss a"))
        .drop("AvailableDtTm")

        fireTsDF
        .select(year(col("IncidentDate")))
        .distinct()
        .orderBy(year(col("IncidentDate")))
        .show()

        fireTsDF
        .select(col("CallType"))
        .where(col("CallType").isNotNull)
        .groupBy(col("CallType"))
        .count()
        .orderBy(desc("count"))
        .show(10, false)

        // What were all the different types of fire calls in 2018?
        fireTsDF
        .select(col("CallType"))
        .where(year(col("IncidentDate"))==="2018")
        .distinct().show(truncate=false)

        // What months within the year 2018 saw the highest number of fire calls?
        fireTsDF
        .select(month(col("IncidentDate")).alias("Month"))
        .where(col("Month").isNotNull)
        .where(year(col("IncidentDate"))==="2018")
        .groupBy(col("Month"))
        .agg(count("*").alias("TotalCalls"))
        .orderBy(desc("TotalCalls"))
        .show(truncate=false)

        // Which neighborhoods had the worst response times to fire calls in 2018?
        fireTsDF
        .select(col("Neighborhood"), col("Delay"))
        .orderBy(desc("Delay"))
        .show()

        fireTsDF
        .select(col("Neighborhood"), col("Delay"))
        .where(year(col("IncidentDate")) === "2018")
        .groupBy(col("Neighborhood"))
        .agg(max(col("Delay")).alias("MaxDelay"))
        .orderBy(desc("MaxDelay"))
        .show(truncate=false)

        val fireCounts = fireDF
        .select(col("Neighborhood"), col("Zipcode"), col("CallType"))
        .groupBy(col("Neighborhood"), col("Zipcode"))
        .agg(count(col("CallType")).alias("FireCallCount"))

        // Encoding neighborhood and zipcode
        val neighborhoodIndexer = new StringIndexer()
        .setInputCol("Neighborhood")
        .setOutputCol("NeighborhoodIndexed")
        .fit(fireCounts)
        .setHandleInvalid("skip")

        val zipcodeIndexer = new StringIndexer()
        .setInputCol("Zipcode")
        .setOutputCol("ZipcodeIndexed")
        .fit(fireCounts)
        .setHandleInvalid("skip")

        val  indexedDF = zipcodeIndexer.transform(neighborhoodIndexer.transform(fireCounts))
        print(indexedDF.printSchema())

        // assemble features vector
        val assembler = new VectorAssembler()
        .setInputCols(Array("NeighborhoodIndexed", "ZipcodeIndexed", "FireCallCount"))
        .setOutputCol("features")

        val finalDf = assembler.transform(indexedDF)

        val Row(coeff: Matrix) = Correlation.corr(finalDf, "features").head()

        print(s"Correlation matrix:\n $coeff")


        println("-----------------------------------------------")

        


        
    }   

    def createLocalFolderIfNotExists(path: String): Unit = {
        val folder = new java.io.File(path)
        if (!folder.exists()){
            val created = folder.mkdirs()
            if (created) println(s"Folder created at : $path")
            else println(s"Failed to create folder at: $path")
        } else {
            println(s"Folder already present at: $path")
        }
    }
}

File: src/main/python/dfops.py
# from pyspark.shell import spark
from pyspark.sql import SparkSession
from pyspark.sql import functions as f
from pyspark.sql.types import (
    StructField,
    StructType,
    IntegerType,
    StringType,
    BooleanType,
    FloatType,
)
from pyspark.ml.feature import StringIndexer
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.stat import Correlation


fire_schema = StructType(
    [
        StructField("CallNumber", IntegerType(), True),
        StructField("UnitID", StringType(), True),
        StructField("IncidentNumber", IntegerType(), True),
        StructField("CallType", StringType(), True),
        StructField("CallDate", StringType(), True),
        StructField("WatchDate", StringType(), True),
        StructField("CallFinalDisposition", StringType(), True),
        StructField("AvailableDtTm", StringType(), True),
        StructField("Address", StringType(), True),
        StructField("City", StringType(), True),
        StructField("Zipcode", IntegerType(), True),
        StructField("Battalion", StringType(), True),
        StructField("StationArea", StringType(), True),
        StructField("Box", StringType(), True),
        StructField("OriginalPriority", StringType(), True),
        StructField("Priority", StringType(), True),
        StructField("FinalPriority", IntegerType(), True),
        StructField("ALSUnit", BooleanType(), True),
        StructField("CallTypeGroup", StringType(), True),
        StructField("NumAlarms", IntegerType(), True),
        StructField("UnitType", StringType(), True),
        StructField("UnitSequenceInCallDispatch", IntegerType(), True),
        StructField("FirePreventionDistrict", StringType(), True),
        StructField("SupervisorDistrict", StringType(), True),
        StructField("Neighborhood", StringType(), True),
        StructField("Location", StringType(), True),
        StructField("RowID", StringType(), True),
        StructField("Delay", FloatType(), True),
    ]
)

spark = SparkSession.builder.appName("FireReader").getOrCreate()

file_loc = "/app/src/main/data/sf-fire-calls.csv"

fire_df = spark.read.schema(fire_schema).option("header", True).csv(file_loc)

parquet_path = "app/src/main/data/python/parquet/fire"


fire_df.write.mode("overwrite").format("parquet").save(parquet_path)
# fire_df.write.mode("overwrite").format("parquet").saveAsTable("fireTable")

few_fire_df = fire_df.select("IncidentNumber", "AvailableDtTm", "CallType").where(
    f.col("CallType") != "Medical Incident"
)

fire_df.select("CallType").where(f.col("CallType").isNotNull()).agg(
    f.count_distinct(f.col("CallType")).alias("DistinctCallType")
).show()

few_fire_df.show(5, truncate=False)

fire_df.select("CallType").where(f.col("CallType").isNotNull()).distinct().show(
    10, truncate=False
)


# renaming column
fire_df.withColumnRenamed("Delay", "ResponseDelayedinMins").select(
    "ResponseDelayedinMins"
).where(f.col("ResponseDelayedinMins") > 5).show()

# converting datetime ts
fire_ts_df = (
    fire_df.withColumn("IncidentDate", f.to_timestamp(f.col("CallDate"), "MM/dd/yyyy"))
    .drop(f.col("CallDate"))
    .withColumn("OnWatchDate", f.to_timestamp(f.col("WatchDate"), "MM/dd/yyyy"))
    .drop(f.col("WatchDate"))
    .withColumn(
        "AvailableDtTs", f.to_timestamp(f.col("AvailableDtTm"), "MM/dd/yyyy hh:mm:ss a")
    )
    .drop(f.col("AvailableDtTm"))
)

fire_ts_df.select("IncidentDate", "OnWatchDate", "AvailableDtTs").show(5, False)


fire_counts = (
    fire_ts_df.select(f.col("Neighborhood"), f.col("Zipcode"))
    .groupBy(f.col("Neighborhood"), f.col("Zipcode"))
    .count()
    .withColumnRenamed("count", "FireCallCounts")
)

# indexing neighborhood and zipcode
neighborhood_indexer = StringIndexer(
    inputCol="Neighborhood", outputCol="NeighborhoodIndexed"
).setHandleInvalid("skip")

zipcode_indexer = StringIndexer(
    inputCol="Zipcode", outputCol="ZipcodeIndexed"
).setHandleInvalid("skip")

indexed_df = neighborhood_indexer.fit(fire_counts).transform(fire_counts)
indexed_df = zipcode_indexer.fit(indexed_df).transform(indexed_df)

# assembling
assembler = VectorAssembler(
    inputCols=["NeighborhoodIndexed", "ZipcodeIndexed", "FireCallCounts"],
    outputCol="features",
).setHandleInvalid("skip")

final_df = assembler.transform(indexed_df)

correlation_matrix = Correlation.corr(final_df, "features").head()[0]

print(correlation_matrix)
